src/com/google/enterprise/adaptor/testing/RecordingDocIdPusher.java

```diff
@@ -12,57 +12,93 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-package com.google.enterprise.adaptor;
+package com.google.enterprise.adaptor.testing;
+
+import static java.util.Collections.unmodifiableList;
+import static java.util.Collections.unmodifiableSet;
+
+import com.google.enterprise.adaptor.AbstractDocIdPusher;
+import com.google.enterprise.adaptor.Acl;
+import com.google.enterprise.adaptor.DocId;
+import com.google.enterprise.adaptor.ExceptionHandler;
+import com.google.enterprise.adaptor.GroupPrincipal;
+import com.google.enterprise.adaptor.Principal;
 
 import java.util.ArrayList;
 import java.util.Collection;
 import java.util.Collections;
+import java.util.LinkedHashSet;
 import java.util.List;
 import java.util.Map;
+import java.util.Set;
 import java.util.TreeMap;
 
-class AccumulatingDocIdPusher extends AbstractDocIdPusher {
+/**
+ * A fake implementation of {@link DocIdPusher} that simply records
+ * the values it receives.
+ */
+public class RecordingDocIdPusher extends AbstractDocIdPusher {
   private List<DocId> ids = new ArrayList<DocId>();
-  private List<DocIdPusher.Record> records
-      = new ArrayList<DocIdPusher.Record>();
+  private List<Record> records = new ArrayList<Record>();
+  private Map<DocId, Acl> namedResources = new TreeMap<DocId, Acl>();
   private Map<GroupPrincipal, Collection<Principal>> groups
       = new TreeMap<GroupPrincipal, Collection<Principal>>();
 
   @Override
   public Record pushRecords(Iterable<Record> records, ExceptionHandler handler)
       throws InterruptedException {
     for (Record record : records) {
-      this.records.add(record);
       ids.add(record.getDocId());
+      this.records.add(record);
     }
     return null;
   }
 
-  public List<DocId> getDocIds() {
-    return Collections.unmodifiableList(ids);
-  }
-
-  public List<DocIdPusher.Record> getRecords() {
-    return Collections.unmodifiableList(records);
-  }
-
-  public void reset() {
-    ids.clear();
-    records.clear();
-  }
-
   @Override
   public DocId pushNamedResources(Map<DocId, Acl> resources,
       ExceptionHandler handler) throws InterruptedException {
-    throw new UnsupportedOperationException();
+    namedResources.putAll(resources);
+    return null;
   }
 
   @Override
   public GroupPrincipal pushGroupDefinitions(
       Map<GroupPrincipal, ? extends Collection<Principal>> defs,
       boolean caseSensitive, ExceptionHandler handler)
       throws InterruptedException {
-    groups.putAll(defs); 
+    for (GroupPrincipal key : defs.keySet()) {
+      Collection<Principal> members = defs.get(key);
+      if (members instanceof List) {
```

Comment:
Most of the tests in the connectors use assertEquals on the collected groups. Most collections are not equal to each other, even if they contain the same members, because they use Object.equals. List and Set are different, but a List only equals another List, and a Set only equals another Set. Some connectors use a Set internally, and others a List. The tests have to use the same type. This code is trying to make a defensive copy, based on Eric's copy in the SharePoint connector. But you can't make a defensive copy that preserves order and equality without knowing whether you are copying a List or a Set.

It is possible that a third-party connector would use another collection type that supports membership equality, but that's rather unlikely, and I just punted on that. (Specifically, Deque and Queue, the other JDK collections, do not support membership equality. Guava has some collections that do support membership equality, such as Multiset and Table. We just won't make a defensive copy of those, but the SharePoint connector was the only one doing that, anyway.)